File: RandomForests/RandomForests.r
library(randomForest)
library(gmodels)

# store csv file in data frame
bank <- read.csv("data/bank.csv", sep=";")

# set random number generator to produce same results for each run
set.seed(123)

# get random set of 900 integers ranging from 1 to 1000 
train_sample <- sample(4521, 3000)

# split the data frames into training and test sets
# -train_sample means all values not in train_sample
bank_train <- bank[train_sample, ]
bank_test  <- bank[-train_sample, ]

# fmla is equivalent to y = x1 + x2 + ..... 
# create decision tree using formula and training set
# setting keep.forest to true means the forest will be saved in the output
fmla <- paste0("y ~ ", paste0(names(bank[-17]), collapse=" + "))
forestModel <- randomForest(formula=as.formula(fmla), data=bank_train, ntree=10000, keep.forest=T)

# get predicted labels for test set
bank_pred <- predict(forestModel, bank_test, type="response")

# create a table of predicted outcomes compared to actual results
# the sum of the entries along the main diagonal are the number of correctly predicted labels
# row, column and chi-squared proportions are not shown
crossTable <- CrossTable(bank_test$y, bank_pred,
						 prop.chisq = FALSE, prop.c = FALSE, prop.r = FALSE,
						 dnn = c('actual y', 'predicted y'))

# get ratio of correctly labeled samples to total samples
print(paste0("prediction accuracy: ", (crossTable$t[1] + crossTable$t[4]) / sum(crossTable$t)))

File: RandomForests/RandomForests.py
from sklearn.ensemble import RandomForestClassifier
from sklearn import preprocessing
import pandas as pd
import random as rm

# store csv file in data frame
bank = pd.read_csv("../data/bank.csv", delimiter=";")

# create a LabelEncoder instance
le = preprocessing.LabelEncoder()

# change categorical data to numerical data for fitting model
# i.e. for bank['y'], no = 0, yes = 1
c = bank.columns.values
i = 0
while i < len(c):
    if bank[c[i]].dtype == 'object':
        le.fit(bank[c[i]])
        bank[c[i]] = le.transform(bank[c[i]])
    i += 1

# set seed value to recreate same random values
rm.seed(123)

# split the data frame into training and test sets
train_sample = rm.sample(range(4521), 3000)
test_sample = [x for x in range(4521) if x not in train_sample]
bank_train = bank.iloc[train_sample,]
bank_test = bank.iloc[test_sample,]

# fit data to random forest
rfc = RandomForestClassifier(n_estimators=10000)
rfc.fit(bank_train.iloc[:, 0:16], bank_train['y'])

# get list of predicted labels for test set
pred_vals = rfc.predict(bank_test.iloc[:, 0:16])

# create a table of predicted outcomes compared to actual results
# the sum of the entries along the main diagonal are the number of correctly predicted labels
print(pd.crosstab(pred_vals, bank_test['y']))

# get ratio of correctly labeled samples to total samples
score = rfc.score(bank_test.iloc[:, 0:16], bank_test['y'])
print("prediction accuracy: ", score)